x/pool-incentives/keeper/grpc_query_test.go

```diff
@@ -150,140 +172,123 @@ func (suite *KeeperTestSuite) TestLockableDurations() {
-
-	suite.Equal(poolId, res.IncentivizedPools[2].PoolId)
-	suite.Equal(gauge3Id, res.IncentivizedPools[2].GaugeId)
-	suite.Equal(time.Hour*7, res.IncentivizedPools[2].LockableDuration)
-
-	suite.Equal(poolId2, res.IncentivizedPools[3].PoolId)
-	suite.Equal(gauge4Id, res.IncentivizedPools[3].GaugeId)
-	suite.Equal(time.Hour*7, res.IncentivizedPools[3].LockableDuration)
-
-	// Ensure that non-perpetual pot can't get rewards.
-	// TODO: extract this to standalone test
-	gauge6Id, err := suite.App.IncentivesKeeper.CreateGauge(
+	gaugeNonPerpetualId, err := suite.App.IncentivesKeeper.CreateGauge(
 		suite.Ctx, notPerpetual, sdk.AccAddress{}, sdk.Coins{}, lockuptypes.QueryCondition{
 			LockQueryType: lockuptypes.ByDuration,
 			Denom:         "stake",
 			Duration:      time.Hour,
 		}, time.Now(), 1)
 
 	suite.NoError(err)
-	err = keeper.UpdateDistrRecords(suite.Ctx, types.DistrRecord{
-		GaugeId: gauge6Id,
-		Weight:  sdk.NewInt(100),
-	})
-	suite.Error(err)
+
+	for _, tc := range []struct {
+		desc                 string
+		poolCreated          bool
+		distRecords          []types.DistrRecord
+		expectedErr          bool
+		expectedRecordLength int
+	}{
+		{
+			desc:                 "No pool exist",
+			poolCreated:          false,
+			distRecords:          []types.DistrRecord{},
+			expectedErr:          false,
+			expectedRecordLength: 0,
+		},
+		{
+			desc:                 "Normal case",
+			poolCreated:          true,
+			distRecords:          []types.DistrRecord{},
+			expectedErr:          false,
+			expectedRecordLength: 3,
+		},
+		{
+			desc:        "Perpetual",
+			poolCreated: true,
+			distRecords: []types.DistrRecord{{
+				GaugeId: gaugePerpetualId,
+				Weight:  sdk.NewInt(300),
+			}},
+			expectedErr:          false,
+			expectedRecordLength: 3,
+		},
+		{
+			desc:        "Non Perpetual",
+			poolCreated: true,
+			distRecords: []types.DistrRecord{{
+				GaugeId: gaugeNonPerpetualId,
+				Weight:  sdk.NewInt(100),
+			}},
+			expectedErr:          true,
+			expectedRecordLength: 0,
+		},
+	} {
+		tc := tc
+		suite.Run(tc.desc, func() {
+			var lockableDurations []time.Duration
+			if tc.poolCreated {
+				// only create a pool
+				pool, err := ammKeeper.GetPoolsAndPoke(suite.Ctx)
+				if len(pool) == 0 {
+					poolId = suite.PrepareBalancerPool()
+				}
+				// LockableDurations should be 1, 3, 7 hours from the default genesis state.
+				lockableDurations = keeper.GetLockableDurations(suite.Ctx)
+				suite.Equal(3, len(lockableDurations))
+
+				var distRecords []types.DistrRecord
+				for i := 0; i < len(lockableDurations); i++ {
+					gaugeId, err := keeper.GetPoolGaugeId(suite.Ctx, poolId, lockableDurations[i])
+					suite.NoError(err)
+
+					distRecords = append(distRecords, types.DistrRecord{
+						GaugeId: gaugeId,
+						Weight:  sdk.NewInt(100),
+					})
+				}
+				tc.distRecords = append(distRecords, tc.distRecords...)
+
+				// Sort in ascending order of gaugeId
```

Comment:
Why do we have put distrRecords in ascending order?

Reply:
I believe the UpdateDistrRecords function only accepts sorted records

Reply:
yeah it errors if not